### DISCUSSION TITLE
Temp+humidity same chip and scan point

### DISCUSSION THREAD

**Discussion by wingspinner** [HomeSpan/HomeSpan#965] [Q&A]
Ive read all the documents and scanned through all the examples as well as the post about multiple sensors on one scan point. The suggested solutions seem non-optimal and haven't clarified for me a clean, simple and efficient way to handle the numerous modern temp chips that also have humidity built in. 

Case in point, I've got a main device and a temp remote device working fine using the SHT40. However the chip also has built in humidity and when I read the SHT40 I get temp and humidity at the same time. How about sending a structure with both pieces of data back to the main device and have it service a sort of virtual accessory for each without having to talk to the remote scan point more than once? Any straightforward way to do this without using globals?

Been doing complex embedded software and hardware for longer than I care to admit however just started looking at HomeKit and HomeSpan in detail a few days ago so I'm quite sure I don't  know what I don't know so please bear with me. 

Thx - Ron

**Comment by HomeSpan** [comment]
You can (and should) use a single structure that reports both temp and humidity - no need to make separate SpanPoint calls back to the Main Device. However, HomeKit itself, as structured by Apple, only supports Accessories that can be organized in relatively fixed ways according to Services.  If you are using a Service that contains multiple sensors, such as a Thermostat (contains both temp and humidity) or an Air Quality Sensor (contains 5 separate measurements), than the multi-value structure can be passed in SpanPoint within that Service.

Unfortunately, there is no single Service in Apple HomeKit that contains just a temperature and humidity sensor, so if you want to use SpanPoint to send that info back in one structure, you'll need to either do it as a global (which for a small embedded system I think is totally fine) and have a process that updates a flag so that each of the Temp and Humidity sensors knows to update its measurements, or you can get creative and find interesting ways of doing this within a custom Service.

I have not tried this myself, but you can create a custom Service that uses SpanPoint (see the docs on how to create custom Services) and have that Service contain two separate sub-services (Temp and Humidity Sensors) to combine everything in one framework.  Similar to how the Faucet Service has a linked Valve Services.

Best thing to do is experiment and find which way works best for you.

**Reply by wingspinner** [reply]
Thanks for the reply. Yes, I'll experiment starting with the custom service. Was hoping to whip this out with minimal experimentation since I've got another much more complex project that needs attention. 

Thanks again

**Reply by ChassaB** [reply]
Well I guess that answers my question. I have an old AM2320 temp and humidity sensor and tried to write a sketch to set up both T and H but got the "accessory out of compliance message". Worked fine when I just set it up as a temp sensor.  Not doing it as a SpanPoint. Just doing some learning by mucking around. Now I'll have to figure out that flag idea.

**Reply by frankonski** [reply]
Works fine for me using a 3-pin DHT22 sensor. I create only one accessory and have a single class that contains both a Temperature service and Humidity service, with necessary Characteristics, where I implemented a `loop()` method that I call from the main loop. From there, I do both readings and update the HS accessory states in a single pass.

I couldn’t figure out how (easily) to have one custom service with two sub-services, as suggested above. Got lazy on this..
